File: scripts/cover_gen.py
#!/usr/bin/env python3
"""Cover image generation via MiniMax image_generation API.

Usage:
  python3 cover_gen.py "标题" ["摘要" ] [output_path] [--model image-01] [--aspect 16:9]

Env vars:
  MINIMAX_API_KEY         MiniMax API key (必填)
  MINIMAX_IMAGE_MODEL     默认 image-01
  MINIMAX_HOST            默认 api.minimaxi.com
  COVER_ASPECT_RATIO      默认 16:9
  COVERS_DIR              默认 ./covers
"""
import argparse
import base64
import http.client
import json
import os
import re
import sys
import time
from pathlib import Path
from urllib.parse import urlparse


def call_image_gen(prompt: str, model: str, aspect_ratio: str, host: str, api_key: str, timeout: int = 120) -> dict:
    body = json.dumps(
        {
            "model": model,
            "prompt": prompt,
            "aspect_ratio": aspect_ratio,
            "n": 1,
            "response_format": "b64_json",
        },
        ensure_ascii=False,
    ).encode("utf-8")
    conn = http.client.HTTPSConnection(host, timeout=timeout)
    conn.request(
        "POST",
        "/v1/image_generation",
        body=body,
        headers={
            "Authorization": f"Bearer {api_key}",
            "x-api-key": api_key,
            "Content-Type": "application/json; charset=utf-8",
        },
    )
    resp = conn.getresponse()
    data = resp.read().decode("utf-8")
    conn.close()
    if resp.status < 200 or resp.status >= 300:
        raise RuntimeError(f"minimax HTTP {resp.status}: {data[:500]}")
    return json.loads(data)


def build_prompt(title: str, description: str = "", style: str = "") -> str:
    style = style or (
        "modern professional editorial, suitable for Chinese WeChat "
        "public account business content, clean composition, soft natural "
        "lighting, subtle palette"
    )
    desc_part = f"\n\nArticle summary: {description}" if description else ""
    return (
        "Generate a cover image for a Chinese WeChat public account article.\n"
        f"Title: {title}{desc_part}\n\n"
        f"Style: {style}.\n"
        "No text, no logos, no watermarks, no people faces in the foreground."
    )


def extract_image(obj: dict) -> tuple[bytes, str, str]:
    """Return (image_bytes, mime, source_label)."""
    candidates = [
        obj.get("data", {}).get("image_base64") if isinstance(obj.get("data"), dict) else None,
        obj.get("image_base64"),
    ]
    data_list = obj.get("data") if isinstance(obj.get("data"), list) else None
    if data_list:
        for item in data_list:
            if isinstance(item, dict):
                candidates.append(item.get("image_base64") or item.get("b64_json"))

    for c in candidates:
        if isinstance(c, str) and len(c) > 200:
            return base64.b64decode(c), "image/jpeg", "b64_json"

    if data_list and isinstance(data_list, list):
        for item in data_list:
            if isinstance(item, dict) and isinstance(item.get("url"), str):
                url = item["url"]
                return _download(url), "image/png" if url.lower().endswith(".png") else "image/jpeg", f"url:{url[:60]}"
    if isinstance(obj.get("data", {}).get("image_url"), str):
        url = obj["data"]["image_url"]
        return _download(url), "image/png" if url.lower().endswith(".png") else "image/jpeg", f"url:{url[:60]}"

    raise RuntimeError(f"no image data in response: {json.dumps(obj)[:500]}")


def _download(url: str) -> bytes:
    import urllib.request
    with urllib.request.urlopen(url, timeout=60) as r:
        return r.read()


def safe_filename(s: str) -> str:
    s = re.sub(r"[^\w\u4e00-\u9fa5_-]", "_", s)
    return s[:60] or "cover"


def generate_cover(
    title: str,
    description: str = "",
    out_path: str | None = None,
    model: str | None = None,
    aspect_ratio: str | None = None,
    covers_dir: str | None = None,
    style: str = "",
) -> dict:
    api_key = os.environ.get("MINIMAX_IMAGE_API_KEY") or os.environ.get("MINIMAX_API_KEY", "")
    if not api_key:
        raise RuntimeError("MINIMAX_API_KEY (or MINIMAX_IMAGE_API_KEY) is not set")
    host = os.environ.get("MINIMAX_HOST", "api.minimaxi.com")
    model = model or os.environ.get("MINIMAX_IMAGE_MODEL", "image-01")
    aspect_ratio = aspect_ratio or os.environ.get("COVER_ASPECT_RATIO", "16:9")
    covers_dir = covers_dir or os.environ.get("COVERS_DIR") or str(Path.cwd() / "covers")

    prompt = build_prompt(title, description, style)
    obj = call_image_gen(prompt, model, aspect_ratio, host, api_key)
    data, mime, source = extract_image(obj)

    Path(covers_dir).mkdir(parents=True, exist_ok=True)
    if not out_path:
        ext = "png" if mime == "image/png" else "jpg"
        out_path = str(Path(covers_dir) / f"{safe_filename(title)}_{int(time.time())}.{ext}")
    Path(out_path).write_bytes(data)
    return {
        "ok": True,
        "file_path": out_path,
        "size": len(data),
        "mime": mime,
        "model": model,
        "source": source,
    }


def main():
    p = argparse.ArgumentParser()
    p.add_argument("title")
    p.add_argument("description", nargs="?", default="")
    p.add_argument("out_path", nargs="?", default=None)
    p.add_argument("--model", default=None)
    p.add_argument("--aspect", dest="aspect_ratio", default=None)
    p.add_argument("--covers-dir", dest="covers_dir", default=None)
    p.add_argument("--style", default="")
    args = p.parse_args()

    result = generate_cover(
        args.title,
        args.description,
        args.out_path,
        args.model,
        args.aspect_ratio,
        args.covers_dir,
        args.style,
    )
    print(json.dumps(result, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(f"ERROR: {e}", file=sys.stderr)
        sys.exit(1)

File: src/cover_gen.ts
import fs from 'fs';
import path from 'path';
import https from 'https';
import crypto from 'crypto';

const MINIMAX_HOST = process.env.MINIMAX_HOST || 'api.minimaxi.com';
const MINIMAX_IMAGE_KEY = process.env.MINIMAX_IMAGE_API_KEY || process.env.MINIMAX_API_KEY || '';
const DEFAULT_MODEL = process.env.MINIMAX_IMAGE_MODEL || 'image-01';
const DEFAULT_ASPECT = process.env.COVER_ASPECT_RATIO || '16:9';
const COVERS_DIR = process.env.COVERS_DIR || path.join(process.cwd(), 'covers');
const DEFAULT_STYLE = process.env.COVER_STYLE || 'auto';
const RETRY_MAX = Math.max(0, parseInt(process.env.COVER_RETRY_MAX || '2', 10));
const RETRY_BASE_MS = Math.max(100, parseInt(process.env.COVER_RETRY_BASE_MS || '1500', 10));
const TEXT_OVERLAY = (process.env.COVER_TEXT_OVERLAY ?? 'true').toLowerCase() !== 'false';

export type CoverStyle = 'auto' | 'tech' | 'finance' | 'lifestyle' | 'education' | 'news';

export interface CoverGenOptions {
  title: string;
  description?: string;
  style?: CoverStyle | string;
  model?: string;
  aspectRatio?: string;
  outPath?: string;
  noText?: boolean;
}

export interface CoverGenResult {
  ok: true;
  filePath: string;
  size: number;
  mime: string;
  model: string;
  style: string;
  attempts: number;
  durationMs: number;
}

export interface CoverGenError {
  ok: false;
  error: string;
  statusCode?: number;
  attempts: number;
}

const STYLE_PRESETS: Record<string, string> = {
  auto: 'modern professional editorial, suitable for Chinese WeChat public account business content, clean composition, soft natural lighting, subtle palette',
  tech: 'futuristic tech aesthetic, deep blue and cyan gradients, abstract geometric shapes and circuit-like patterns, soft glow, no human faces, suitable for AI / programming / software topics',
  finance: 'business finance aesthetic, dark navy with gold accents, abstract candlestick and chart silhouettes, upward arrows, professional and authoritative, no human faces, suitable for stocks / economy / markets topics',
  lifestyle: 'lifestyle aesthetic, warm pastel tones, soft natural daylight, cozy lifestyle objects (coffee, plants, books, window light), relaxed and inviting, no human faces, suitable for personal growth / wellness topics',
  education: 'educational aesthetic, light cream background with subtle paper texture, soft bookshelf or chalkboard motifs, warm and intellectual, no human faces, suitable for learning / knowledge / tutorial topics',
  news: 'news editorial aesthetic, bold red accent on neutral grayscale, sharp geometric shapes reminiscent of broadcast graphics, urgent and serious tone, no human faces, suitable for current events / breaking news topics',
};

const VALID_STYLES: CoverStyle[] = ['auto', 'tech', 'finance', 'lifestyle', 'education', 'news'];

function resolveStyle(style?: string): string {
  const s = (style || DEFAULT_STYLE || 'auto').toLowerCase();
  if (!VALID_STYLES.includes(s as CoverStyle)) {
    console.warn(`[cover-gen] unknown style '${s}', falling back to 'auto'`);
    return STYLE_PRESETS.auto;
  }
  return STYLE_PRESETS[s];
}

function buildPrompt(opts: CoverGenOptions): string {
  const styleDesc = resolveStyle(opts.style);
  const descPart = opts.description ? `\n\nArticle summary: ${opts.description}` : '';
  const overlay = !opts.noText && TEXT_OVERLAY;
  const textDirective = overlay
    ? `\n\nRender the article title prominently as bold elegant Chinese typography. Title text to render:\n"""${opts.title}"""\nUse a clear readable font, high contrast against the background, with subtle shadow or highlight for legibility. The title should occupy the central area of the image, scaled to be readable when displayed as a small thumbnail.`
    : `\n\nDo NOT render any text, logos, watermarks, or recognizable human faces.`;

  return `Generate a cover image for a Chinese WeChat public account article.
Title: ${opts.title}${descPart}

Visual style: ${styleDesc}.${textDirective}

Technical requirements:
- Aspect ratio: ${opts.aspectRatio || DEFAULT_ASPECT} (WeChat list-card friendly)
- No other text besides the title
- No watermarks, signatures, or copyright marks
- No recognizable human faces in the foreground`;
}

function postJson(host: string, urlPath: string, headers: Record<string, string>, body: any, timeoutMs = 90000): Promise<{ status: number; data: string }> {
  return new Promise((resolve, reject) => {
    const payload = Buffer.from(JSON.stringify(body), 'utf-8');
    const req = https.request({
      host,
      port: 443,
      path: urlPath,
      method: 'POST',
      headers: { ...headers, 'Content-Length': payload.length.toString() },
      timeout: timeoutMs,
    }, (res) => {
      const chunks: Buffer[] = [];
      res.on('data', (c) => chunks.push(c));
      res.on('end', () => resolve({ status: res.statusCode || 0, data: Buffer.concat(chunks).toString('utf-8') }));
    });
    req.on('timeout', () => { req.destroy(new Error('request timeout')); });
    req.on('error', reject);
    req.write(payload);
    req.end();
  });
}

function extractBase64DataUrl(respBody: string): { mime: string; base64: string; downloadUrl?: string } | null {
  let obj: any;
  try { obj = JSON.parse(respBody); } catch { return null; }

  const candidates: any[] = [
    obj?.data?.image_base64,
    obj?.image_base64,
    obj?.data?.[0]?.image_base64,
    obj?.data?.[0]?.b64_json,
    obj?.b64_json,
  ];
  for (const c of candidates) {
    if (typeof c === 'string' && c.length > 100) {
      return { mime: 'image/jpeg', base64: c };
    }
  }
  const urlCandidate = obj?.data?.image_url || obj?.data?.[0]?.url || obj?.url;
  if (typeof urlCandidate === 'string' && urlCandidate.startsWith('http')) {
    return { mime: '', base64: '', downloadUrl: urlCandidate };
  }
  return null;
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

function isTransient(status: number): boolean {
  return status === 429 || (status >= 500 && status < 600);
}

export async function generateCover(opts: CoverGenOptions): Promise<CoverGenResult | CoverGenError> {
  if (!MINIMAX_IMAGE_KEY) {
    return { ok: false, error: 'MINIMAX_IMAGE_API_KEY (or MINIMAX_API_KEY) is not set', attempts: 0 };
  }
  const prompt = buildPrompt(opts);
  const model = opts.model || DEFAULT_MODEL;
  const aspectRatio = opts.aspectRatio || DEFAULT_ASPECT;
  const styleName = (opts.style || DEFAULT_STYLE || 'auto').toLowerCase();

  const body = {
    model,
    prompt,
    aspect_ratio: aspectRatio,
    n: 1,
    response_format: 'b64_json',
  };

  if (!fs.existsSync(COVERS_DIR)) {
    fs.mkdirSync(COVERS_DIR, { recursive: true });
  }

  const titleHash = crypto.createHash('sha1').update(`${styleName}|${opts.title}|${opts.description || ''}`).digest('hex').slice(0, 10);
  const safeTitle = (opts.title || 'cover').replace(/[^a-zA-Z0-9\u4e00-\u9fa5_-]/g, '_').slice(0, 30);
  const ts = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
  const filePath = opts.outPath || path.join(COVERS_DIR, `${safeTitle}_${titleHash}_${ts}.jpg`);

  let lastErr = '';
  let lastStatus: number | undefined;
  const startedAt = Date.now();

  for (let attempt = 1; attempt <= RETRY_MAX + 1; attempt++) {
    let resp;
    try {
      resp = await postJson(MINIMAX_HOST, '/v1/image_generation', {
        'Authorization': `Bearer ${MINIMAX_IMAGE_KEY}`,
        'x-api-key': MINIMAX_IMAGE_KEY,
        'Content-Type': 'application/json; charset=utf-8',
      }, body);
    } catch (e: any) {
      lastErr = `minimax request failed: ${e?.message || e}`;
      console.warn(`[cover-gen] attempt ${attempt} failed: ${lastErr}`);
      if (attempt > RETRY_MAX) break;
      await sleep(RETRY_BASE_MS * Math.pow(2, attempt - 1));
      continue;
    }

    lastStatus = resp.status;
    if (resp.status < 200 || resp.status >= 300) {
      lastErr = `minimax HTTP ${resp.status}: ${resp.data.slice(0, 500)}`;
      console.warn(`[cover-gen] attempt ${attempt} ${lastErr}`);
      if (!isTransient(resp.status) || attempt > RETRY_MAX) break;
      await sleep(RETRY_BASE_MS * Math.pow(2, attempt - 1));
      continue;
    }

    const extracted = extractBase64DataUrl(resp.data);
    if (!extracted) {
      lastErr = `unexpected response shape: ${resp.data.slice(0, 500)}`;
      console.warn(`[cover-gen] attempt ${attempt} ${lastErr}`);
      if (attempt > RETRY_MAX) break;
      await sleep(RETRY_BASE_MS * Math.pow(2, attempt - 1));
      continue;
    }

    let buf: Buffer;
    try {
      if (extracted.downloadUrl) {
        buf = await downloadToBuffer(extracted.downloadUrl);
      } else {
        buf = Buffer.from(extracted.base64, 'base64');
      }
    } catch (e: any) {
      lastErr = `download/decode failed: ${e?.message || e}`;
      console.warn(`[cover-gen] attempt ${attempt} ${lastErr}`);
      if (attempt > RETRY_MAX) break;
      await sleep(RETRY_BASE_MS * Math.pow(2, attempt - 1));
      continue;
    }

    if (buf.length < 1024) {
      lastErr = `image too small (${buf.length} bytes), likely invalid`;
      console.warn(`[cover-gen] attempt ${attempt} ${lastErr}`);
      if (attempt > RETRY_MAX) break;
      await sleep(RETRY_BASE_MS * Math.pow(2, attempt - 1));
      continue;
    }

    const mime = extracted.mime || 'image/jpeg';
    fs.writeFileSync(filePath, buf);
    return {
      ok: true,
      filePath,
      size: buf.length,
      mime,
      model,
      style: styleName,
      attempts: attempt,
      durationMs: Date.now() - startedAt,
    };
  }

  return {
    ok: false,
    error: lastErr || 'cover generation failed after retries',
    statusCode: lastStatus,
    attempts: RETRY_MAX + 1,
  };
}

function downloadToBuffer(url: string): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    https.get(url, (res) => {
      if (res.statusCode && res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        return downloadToBuffer(res.headers.location).then(resolve, reject);
      }
      const chunks: Buffer[] = [];
      res.on('data', (c) => chunks.push(c));
      res.on('end', () => resolve(Buffer.concat(chunks)));
      res.on('error', reject);
    }).on('error', reject);
  });
}

export const __testing = { STYLE_PRESETS, VALID_STYLES, resolveStyle, buildPrompt, isTransient };